common/lib/plugins/failover2/failover2_plugin.ts

```diff
@@ -0,0 +1,575 @@
+import { Messages } from "../../utils/messages";
+import { HostChangeOptions } from "../../host_change_options";
+import { OldConnectionSuggestionAction } from "../../old_connection_suggestion_action";
+import {
+  AwsWrapperError,
+  FailoverFailedError,
+  FailoverSuccessError,
+  TransactionResolutionUnknownError,
+  UnavailableHostError
+} from "../../utils/errors";
+import { getWriter, shuffleList } from "../../utils/utils";
+import { ClientWrapper } from "../../client_wrapper";
+import { HostAvailability } from "../../host_availability/host_availability";
+import { TelemetryTraceLevel } from "../../utils/telemetry/telemetry_trace_level";
+import { HostRole } from "../../host_role";
+import { SubscribedMethodHelper } from "../../utils/subscribed_method_helper";
+
+export class Failover2Plugin extends AbstractConnectionPlugin {
+  private static readonly TELEMETRY_WRITER_FAILOVER = "failover to writer instance";
+  private static readonly TELEMETRY_READER_FAILOVER = "failover to replica";
+  private static readonly METHOD_END = "end";
+  private static readonly INTERNAL_CONNECT_PROPERTY_NAME: string = "76c06979-49c4-4c86-9600-a63605b83f50";
+  private static readonly subscribedMethods: Set<string> = new Set([
+    "initHostProvider",
+    "connect",
+    "forceConnect",
+    "query",
+    "notifyConnectionChanged",
+    "notifyHostListChanged"
+  ]);
+  private readonly _staleDnsHelper: StaleDnsHelper;
+  private readonly _properties: Map<string, any>;
+  private pluginService: PluginService;
+  private readonly _rdsHelper: RdsUtils;
+  private readonly failoverWriterTriggeredCounter: TelemetryCounter;
+  private readonly failoverWriterSuccessCounter: TelemetryCounter;
+  private readonly failoverWriterFailedCounter: TelemetryCounter;
+  private readonly failoverReaderTriggeredCounter: TelemetryCounter;
+  private readonly failoverReaderSuccessCounter: TelemetryCounter;
+  private readonly failoverReaderFailedCounter: TelemetryCounter;
+  private telemetryFailoverAdditionalTopTraceSetting: boolean = false;
+  private _rdsUrlType: RdsUrlType | null = null;
+  private _isInTransaction: boolean = false;
+  private _closedExplicitly: boolean = false;
+  private _lastError: any;
+  /*protected failoverClusterTopologyRefreshRateMsSetting: number = WrapperProperties.FAILOVER_CLUSTER_TOPOLOGY_REFRESH_RATE_MS.defaultValue;
+  protected failoverWriterReconnectIntervalMsSetting: number = WrapperProperties.FAILOVER_WRITER_RECONNECT_INTERVAL_MS.defaultValue;
+  protected failoverReaderConnectTimeoutMsSetting: number = WrapperProperties.FAILOVER_READER_CONNECT_TIMEOUT_MS.defaultValue;*/
+  protected isClosed: boolean = false;
+  failoverMode: FailoverMode | null = null;
+
+  private hostListProviderService?: HostListProviderService;
+  protected enableFailoverSetting: boolean = WrapperProperties.ENABLE_CLUSTER_AWARE_FAILOVER.defaultValue;
+  private failoverTimeoutSettingMs: number = WrapperProperties.FAILOVER_TIMEOUT_MS.defaultValue;
```

Comment:
```suggestion
  private readonly failoverTimeoutSettingMs: number = WrapperProperties.FAILOVER_TIMEOUT_MS.defaultValue;
```

Reply:
If we are assigning them immediately in the constructor what is the point of setting default values here?
```
  protected enableFailoverSetting: boolean;
  private readonly failoverTimeoutSettingMs: number;
```